### PR TITLE
Support new Copilot tool

### PR DESCRIPTION
**Pull request by ecourtois**
Add support for Copilot.
Exchange Protobuf messages between mcu-firmware and Copilot using uartpb module.

### CODE REVIEW

File: lib/cogip_defs/PB_Polar.proto [hunk 1]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
```
Comment (gdoffe) [inline lib/cogip_defs/PB_Polar.proto:1]: > add corresponding Protobuf message

Corresponding to what ?

File: lib/obstacles/include/obstacles/List.hpp [hunk 2]
```diff
@@ -23,13 +23,21 @@
 #include "obstacles/Obstacle.hpp"
 #include "tracefd/File.hpp"
 
+#include "PB_Obstacle.hpp"
+
+
+#define OBSTACLES_MAX_NUMBER 32
```
Comment (gdoffe) [inline lib/obstacles/include/obstacles/List.hpp:29]: This macro is useful but this parameter should be changeable from cmdline.
Thus I would use something like:
```C
#if !defined(OBSTACLES_MAX_NUMBER)
#define OBSTACLES_MAX_NUMBER 32
#endif
```

File: lib/avoidance/include/avoidance.hpp [hunk 3]
```diff
@@ -35,12 +35,15 @@
...
+
 /**
  * @name Dijkstra avoidance parameters
  * @{
  */
-#define GRAPH_MAX_VERTICES      64          /**< Maximum graph vertice number */
-#define DIJKSTRA_MAX_DISTANCE   UINT32_MAX  /**< Maximum Dijkstra distance */
+#define GRAPH_MAX_VERTICES         64          /**< Maximum graph vertice number */
+#define DIJKSTRA_MAX_DISTANCE      UINT32_MAX  /**< Maximum Dijkstra distance */
+#define AVOIDANCE_MAX_POINTS_PATH  32          /**< Maximum number of points in avoidance path */
```
Comment (gdoffe) [inline lib/avoidance/include/avoidance.hpp:46]: What is the purpose of AVOIDANCE_MAX_POINTS_PATH ???
And why 32 ?

File: lib/avoidance/include/avoidance.hpp [hunk 4]
```diff
@@ -84,4 +87,15 @@ bool avoidance_check_recompute(const cogip::cogip_defs::Coords &start,
  */
 void avoidance_print_path(cogip::tracefd::File &out);
 
+/**
+ * @brief Copy list of intermediate positions to reach wanted path positio
```
Comment (gdoffe) [inline lib/avoidance/include/avoidance.hpp:91]: s/positio/position/g

File: platforms/pf_test/platform.cpp [hunk 5]
```diff
@@ -284,6 +294,23 @@ void pf_init_tasks(void)
...
+    /* Initialize UARTPB */
+    uartpb = new cogip::uartpb::UartProtobuf(
+        message_handler,
+        UART_DEV(1)
+        );
+
+    bool uartpb_res = uartpb->connect();
+    if (! uartpb_res) {
+        uartpb = nullptr;
+        cogip::tracefd::out.logf("UART initialization status: %d\n", uartpb_res);
```
Comment (gdoffe) [inline platforms/pf_test/platform.cpp:440]: "UART initialization failed, error: %d\n"

File: platforms/pf_test/platform.cpp [hunk 6]
```diff
@@ -32,6 +34,9 @@
 #define ENABLE_DEBUG        (0)
 #include "debug.h"
 
+/// Maximum number of arguments to shell command callbacks
+#define MAX_COMMAND_ARGS 16
```
Comment (gdoffe) [inline platforms/pf_test/platform.cpp:39]: 16 is not too much ? 8 should be enough ?

File: platforms/pf_test/platform.cpp [hunk 7]
```diff
@@ -249,6 +258,87 @@ static void *_task_planner_start_cancel(void *arg)
...
+    // Computed number of arguments
+    int argc = 0;
+    // List of arguments null-separated
+    char args[COMMAND_DESC_MAX_LENGTH];
+    // Array of pointers to each argument
+    char *argv[MAX_COMMAND_ARGS];
+    // First argument is the command
+    argv[argc++] = (char *)pb_command->cmd();
+    if (pb_command->get_desc().get_length() != 0) {
+        // If there are arguments to the command in the 'desc' attribute
```
Comment (gdoffe) [inline platforms/pf_test/platform.cpp:299]: // If there are arguments **to pass to** the command in the 'desc' attribute

File: platforms/pf_test/platform.cpp [hunk 8]
```diff
@@ -249,6 +258,87 @@ static void *_task_planner_start_cancel(void *arg)
...
+        size_t i;
+        // Copy first argument pointer to 'argv'
+        argv[argc++] = args;
+        for (i = 0; i < pb_command->get_desc().get_length(); i++) {
+            assert(i < COMMAND_DESC_MAX_LENGTH);
+            assert(argc < MAX_COMMAND_ARGS);
```
Comment (gdoffe) [inline platforms/pf_test/platform.cpp:305]: assert are not a bit overkill in such case ?

File: platforms/pf_test/platform.cpp [hunk 9]
```diff
@@ -249,6 +258,87 @@ static void *_task_planner_start_cancel(void *arg)
...
+            // Copy each argument in 'args'
+            args[i] = c;
+            if (c == ' ') {
+                // Insert a null character between each argument
+                args[i] = '\0';
+                // Copy next argument pointer to 'argv'
+                argv[argc++] = args + i + 1;
+            }
+        }
+        // Add null-character afeter last argument
```
Comment (gdoffe) [inline platforms/pf_test/platform.cpp:316]: s/afeter/after

**Review by gdoffe — CHANGES_REQUESTED** [review]

File: platforms/pf_test/platform.cpp [hunk 10]
```diff
@@ -249,6 +258,87 @@ static void *_task_planner_start_cancel(void *arg)
...
+    for (auto command: cogip::shell::global_commands) {
+        if (command->name() == pb_command->cmd()) {
+            run_command(command, pb_command);
+            return;
+        }
+    }
+}
+
+// Read incoming Protobuf message and call the corresponding message handler
+void message_handler(uint8_t message_type, cogip::uartpb::ReadBuffer &buffer)
```
Comment (gdoffe) [inline platforms/pf_test/platform.cpp:345]: This function is used where ? No prototype in a header ?

File: platforms/cortex/trace_utils.cpp [hunk 11]
```diff
@@ -5,20 +5,13 @@
...
-
-/* Trace file descriptor on sdcard */
-static cogip::tracefd::File *tracefd_sdcard;
-
 /* Periodic task */
 #define TASK_PERIOD_MS 60
 
 /* Thread stack */
-static char trace_thread_stack[THREAD_STACKSIZE_LARGE];
+static char trace_thread_stack[THREAD_STACKSIZE_MEDIUM];
```
[the remark this answers is not in the document]
Comment (ecourtois) [reply platforms/cortex/trace_utils.cpp:14]: Yes, it is tested, there was no reason to use LARGE, MEDIUM is enough for the new version of this thread.

File: sys/uartpb/UartProtobuf.cpp [hunk 12]
```diff
@@ -81,7 +81,9 @@ void UartProtobuf::message_reader()
             uint8_t c = ringbuffer_get_one(&rx_buf_);
```
[the remark this answers is not in the document]
Comment (ecourtois) [reply sys/uartpb/UartProtobuf.cpp:90]: Done

File: platforms/cortex/platform.cpp [hunk 13]
```diff
@@ -305,9 +450,25 @@ void pf_init(void)
...
+    /* Initialize UARTPB */
+    uartpb = new cogip::uartpb::UartProtobuf(
+        message_handler,
+        UART_DEV(1)
+        );
+
+    bool uartpb_res = uartpb->connect();
+    if (! uartpb_res) {
+        uartpb = nullptr;
+        cogip::tracefd::out.logf("UART initialization status: %d\n", uartpb_res);
```
Comment (gdoffe) [inline platforms/cortex/platform.cpp:464]: "UART initialization failed, error: %d\n"

**Review by gdoffe — CHANGES_REQUESTED** [review]

File: lib/cogip_defs/include/cogip_defs/Coords.hpp [hunk 14]
```diff
@@ -24,6 +26,9 @@ class Coords {
         double y=0.0        ///< [in] Y coordinate
         ) : x_(x), y_(y) {};
 
+    /// Constructor from Protobuf class
```
[the remark this answers is not in the document]
Comment (ecourtois) [reply lib/cogip_defs/include/cogip_defs/Coords.hpp:29]: When the firmware sends the robot state, it includes pose_current and pose_order, and class Pose inherits class Coords.

File: lib/cogip_defs/PB_Coords.proto [hunk 15]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
```
[the remark this answers is not in the document]
Comment (ecourtois) [reply lib/cogip_defs/PB_Coords.proto:1]: Changed to "add Protobuf message corresponding to class Coords"

File: platforms/pf_test/platform.cpp (hunk 5 above)
Comment (ecourtois) [reply platforms/pf_test/platform.cpp:440]: Done, for cortex platform too.
Commit message for cortex platform was wrong too, also fixed (cup2021 => cortex).

File: platforms/cortex/platform.cpp (hunk 13 above)
Comment (ecourtois) [reply platforms/cortex/platform.cpp:464]: Done

File: lib/obstacles/include/obstacles/List.hpp (hunk 2 above)
Comment (ecourtois) [reply lib/obstacles/include/obstacles/List.hpp:29]: Done

File: lib/avoidance/include/avoidance.hpp (hunk 4 above)
Comment (ecourtois) [reply lib/avoidance/include/avoidance.hpp:91]: Done

File: platforms/pf_test/platform.cpp (hunk 6 above)
Comment (ecourtois) [reply platforms/pf_test/platform.cpp:39]: Yes, 8 is enough.

File: platforms/pf_test/platform.cpp (hunk 7 above)
Comment (ecourtois) [reply platforms/pf_test/platform.cpp:299]: Done

File: platforms/pf_test/platform.cpp (hunk 9 above)
Comment (ecourtois) [reply platforms/pf_test/platform.cpp:316]: Done

File: lib/avoidance/include/avoidance.hpp (hunk 3 above)
Comment (ecourtois) [reply lib/avoidance/include/avoidance.hpp:46]: Used GRAPH_MAX_VERTICES instead.

File: platforms/pf_test/platform.cpp (hunk 10 above)
Comment (ecourtois) [reply platforms/pf_test/platform.cpp:345]: It is a parameter of UartProtobuf constructor.

File: platforms/pf_test/platform.cpp (hunk 8 above)
Comment (ecourtois) [reply platforms/pf_test/platform.cpp:305]: Yes, now I skip the command if its too long, and print a message.

**Review by gdoffe — APPROVED** [review]

**Comment by gdoffe** [comment]
Huge and nice PR ! Good job !!!

**Pull request merged**